### PR TITLE
Update SQL backend to support SQLAlchemy 2.x

### PR DESCRIPTION
**Pull request by davidmezzetti**
Hi @j6k4m8

Once again thank you for this compelling series of graph-related libraries!

I am working to use `grand` as a `NetworkX` compatible backend with `Postgres`. The main issue I ran into is compatibility with SQLAlchemy 2.x. This PR adds support for that. 

This PR also adds a few performance improvements.

- New batch insert method for nodes and edges. A default implementation is provided for other backends.
- Index on edges source and target. This helps with large graphs stored in the database.

A couple other performance improvements were possible using what's already provided in the library (i.e. caching). So overall, with these changes, I'm add to get solid performance with large graphs stored in Postgres.

### CODE REVIEW

File: grand/backends/_sqlbackend.py [hunk 1]
```diff
@@ -60,51 +60,48 @@ def __init__(
...
+            self._node_table_name,
+            self._metadata,
+            sqlalchemy.Column(
+                self._primary_key,
+                sqlalchemy.String(_DEFAULT_SQL_STR_LEN),
+                primary_key=True,
+            ),
+            sqlalchemy.Column("_metadata", sqlalchemy.JSON),
+        )
+        self._node_table.create(self._engine, checkfirst=True)
```
Comment (j6k4m8) [inline grand/backends/_sqlbackend.py:74]: Didn't know about this `checkfirst` arg, that's way more elegant!

File: grand/backends/_sqlbackend.py [hunk 2]
```diff
@@ -60,51 +60,48 @@ def __init__(
...
+                primary_key=True,
+            ),
+            sqlalchemy.Column("_metadata", sqlalchemy.JSON),
+            source_column,
+            target_column
+        )
+        self._edge_table.create(self._engine, checkfirst=True)
+
+        # Create source and target index
+        sindex = Index("edge_source", source_column)
```
Comment (j6k4m8) [inline grand/backends/_sqlbackend.py:100]: 👌

**Review by j6k4m8 — COMMENTED** [review]
Wow @davidmezzetti this is really great stuff! Thank you for the great improvements here!

Running some basic workloads on this today and will merge asap!

**Comment by j6k4m8** [comment]
https://github.com/aplbrain/grand/actions/runs/8649244221/job/23717215080?pr=45#step:6:22

@davidmezzetti Some failing tests — looks like it's a matter of deduplicating edges?

**Comment by davidmezzetti** [comment]
Let me take a look. I didn't see that tests were available. I'll reproduce locally and push a new commit with the fix.

**Comment by j6k4m8** [comment]
Awesome, let me know if you run into any trouble! SHOULD be as simple as running `pytest` (and [it should know to run sql queries in-memory during testing](https://github.com/aplbrain/grand/blob/master/grand/backends/test_backends.py#L69-L78), etc)

**Comment by davidmezzetti** [comment]
Ok, I just checked in a change that fixes the unit tests. One of the errors didn't properly convert the logic to SQLAlchemy 2.x.

The other one is related to the new edge source and target indexes. Previously, when that index wasn't there, the unit tests used the default sort order of the table when pulling edges, which typically uses the primary key. An order_by clause was added to sort by the primary key. 

Alternatively, the [`bfs_successors` calls](https://github.com/aplbrain/grand/blob/master/grand/backends/test_backends.py#L249) could have a `sort_neighbors` argument applied in the tests to guarantee a consistent ordering. This might be a better way to ensure the test always works regardless of the edge ordering.

**Comment by davidmezzetti** [comment]
Looks like this failure is due to not being able to build `networkit` for 3.7. Perhaps 3.7 should be removed given it's long been EOL?

**Comment by j6k4m8** [comment]
My thoughts exactly — want to remove 3.7 and add 3.11?

**Comment by davidmezzetti** [comment]
Just made that change to the build script.

**Comment by j6k4m8** [comment]
Whew — thank you!! Looking good, good to merge once these pass?

**Comment by davidmezzetti** [comment]
Good from my end. 

I'm putting together an example project for txtai that shows how to store all components in Postgres. This will help with the Graph piece and it's really cool how this integrates with the previous work with openCypher!

**Comment by davidmezzetti** [comment]
Hi @j6k4m8 - I wanted to check and see if there were plans to push a new release anytime soon. I'm leaning towards integrating the database graph component as a core feature but would need this change to add the dependency.

**Comment by j6k4m8** [comment]
Sorry, have been traveling — will roll it today!

**Comment by davidmezzetti** [comment]
Great, thank you!

**Pull request merged**